src/backend/distributed/executor/multi_server_executor.c

```diff
@@ -127,8 +131,14 @@ RouterExecutablePlan(MultiPlan *multiPlan, MultiExecutorType executorType)
 	List *workerDependentTaskList = NIL;
 	bool masterQueryHasAggregates = false;
 
-	/* router executor cannot execute queries that hit more than one shard */
-	if (taskCount != 1)
+	/* allow all INSERT ... SELECT queries (i.e., irrespective of taskCount) */
```

Comment:
@anarazel This part is where I'm not very confident.

To switch to router executor, I proposed such a hack(?). I'm open to any suggestions. 
I considered adding `insertSelectJob` boolean to the workerJob. Do you think would that be overkill for this?